thanosql/_util.py

```diff
@@ -0,0 +1,145 @@
+import json
+from typing import Iterable, List, Optional, Tuple, Union
+
+from jinja2 import BaseLoader, Environment, StrictUndefined
+
+from thanosql._error import ThanoSQLValueError
+
+
+def _stringify_list(val: object) -> str:
+    # If we use str(lst) to show the string representation of a list directly,
+    # there can be unwanted extra quotes. e.g. ['a', 'b'] -> ["'a'", "'b'"]
+    # so we have to join list elements manually; this also covers nested lists
+    if isinstance(val, list):
+        return f"[{', '.join([_stringify_list(item) for item in val])}]"
+    return str(val)
+
+
+def _to_postgresql_value_helper(val: object) -> Union[str, list]:
+    # Helper function to convert Python objects into their PSQL representation
+    # First, recursively apply the function to each element of lists
+    if isinstance(val, list):
+        return [_to_postgresql_value_helper(item) for item in val]
+
+    # Convert Python None values to PSQL NULL (without single quotes)
+    if val is None:
+        return "NULL"
+
+    # If the object is a dictionary, convert it to string since we are going
+    # to use its string representation in our PSQL query
+    if isinstance(val, dict):
+        val = json.dumps(val)
+
+    # If the object is a string (including the stringified dictionary above),
+    # replace quoted single quotes (') from \' to '' (C-style to PSQL-style)
+    # Double quotes (") need not be replaced as they are treated as a normal
+    # character in PSQL strings, and are also C-style escaped in JSONs
+    # Then, enclose string/varchar values in single quotes (this includes
+    # JSON/dictionary objects)
+    if isinstance(val, str):
+        quoted_val = val.replace("\\'", "'").replace("'", "''")
+        return f"'{quoted_val}'"
+
+    # For other types of objects, return their string representation
+    return str(val)
+
+
+def _to_postgresql_value(val: object) -> str:
+    val = _to_postgresql_value_helper(val)
+    # There are two possible array representations in PSQL,
+    # '{{...}, {...}}' and ARRAY[[...], [...]]
+    # For empty arrays/lists, we will use the '{}' notation
+    # For non-empty arrays/lists, we will use the ARRAY[[...]] notation
+    if isinstance(val, list):
+        if len(val) == 0:
+            return "'{}'"
+        return f"ARRAY{_stringify_list(val)}"
+    # All other values do not need special treatment
+    return val
+
+
+def _split_query(query: str) -> Tuple[str, str]:
+    # Make sure there is only one {{ val }} placeholder
```

Comment:
If it is simply for split format, can't it be used as '%s' like pyscopg?

Reply:
Technically we can, but we figured it would be better to use jinja format as well since it fits the entire theme better. I think it's a little bit off if query templates use jinja formatting but suddenly execute_values is the only one using "%s" (while its template also uses jinja formatting).

Reply:
{{ val }} o
{{val}}, {{ varibale }}, {{ var }} x
In my opinion, it feels neither intuitive nor flexible in terms of usability.

Reply:
what if we allow both {{val}} and {{ val }} (or anything in between ex. {{ val}})? Would that be more intuitive in your opinion? Or it can be any variable name as long as it is in between {{}}.

Reply:
아니면 since, we are inserting values maybe we should name it {{ values }}.